text/0006-monolothic-packaging.md

```diff
@@ -0,0 +1,343 @@
++-- leftpad (1.0)
+```
+
+Loading dependencies is done in NPM by calling  require("leftpad") or import { ... } from "leftpad". If this statement is executed in a source file of *app*, it will load *leftpad@1.0*, and if it is executed in a source file of *tablify* it will load *leftpad@2.0*.
+
+> The feature of multiple package copies that get loaded in a context-sensitive way is not a common one. The only other package manager that I know of that supports this is Rust’s package manager, Cargo. Most other package managers require all packages to occur a single time in the dependency closure. To do so, I believe most employ a conflict resolution mechanism that boils down to “version specification closest to the root wins” (feel free to add counterexamples to an appendix if I got this wrong).
+### The Problem
+
+All is well with this strategy of keeping multiple copies of a package in the dependency tree, as long as no types “escape” from the boundary of the package’s mini-closure. In the typical case of *tablify* and *leftpad*, [one can reasonably assume that] all values going in and out of the libraries are types from the standard, shared runtime (in this case *strings*), and the only thing being depended upon is a specific behavior provided by the library.
+
+However, let’s say that *leftpad* contains a class called Paddable that *tablify* accepts in its public interface, and that the interface of the class had undergone a breaking change between version 1 and 2 (in JavaScript):
+
+```ts
+// --------------- leftpad 2.0 ---------------------------
+class Paddable {
+  constructor(s) { ... }
+  padLeft(n) { ... } 
+}
+// --------------- tablify 1.0 ---------------------------
+// Expected: Array<Array<Paddable>>
+function tablify(rows) {
+  return rows.map(row => row.map(cell => cell.padLeft(10));
+}
+// --------------- leftpad 1.0 ---------------------------
+class Paddable {
+  constructor(s) { ... }
+  
+  // Oops, forgot to camelCase, rectified in version 2.0
+  padleft(n) { ... } 
+}
+// --------------- app 1.0 ---------------------------
+import { Paddable } from 'leftpad';
+import { tablify } from 'tablify';
+const p = new Paddable('text');
+tablify([[p]]);
+```
+In this code sample, a *Paddable@1* gets constructed and passed to a function which expects to be able to use it as a *Paddable@2*, and the code explodes. 
+Every individual package here was locally correct, but the combination of package versions didn’t work.
+### TypeScript to the rescue
+The CDK uses TypeScript, and the TypeScript compiler actually protects against this issue. If we actually typed the *tablify* function as
+```ts
+import { Paddable } from 'leftpad';
+function tablify(rows:* *Array<Array<Paddable>>) { 
+  ...
+}
+```
+The TypeScript compiler would correctly resolve that type to Array<Array<Paddable@2>>, and refuse to compile the calling of *tablify* with an argument of type Array<Array<Paddable@1>>, because of the detected incompatibility between the types.
```

Comment:
```suggestion
The TypeScript compiler would correctly resolve that type to `Array<Array<Paddable@2>>`, and refuse to compile the calling of *tablify* with an argument of type `Array<Array<Paddable@1>>`, because of the detected incompatibility between the types.
```